src/NuGet.Core/NuGet.Protocol/Utility/DedicatedAsynchronousProcessingThread.cs

```
@@ -0,0 +1,115 @@
+    /// </summary>
+    internal class DedicatedAsynchronousProcessingThread : IDisposable
+    {
+        private Task _processingThread;
+        private bool _isDisposed;
+        private bool _isClosed;
+        private readonly int _pollingDelayInMilliseconds;
+        private readonly ConcurrentQueue<Func<Task>> _taskQueue = new ConcurrentQueue<Func<Task>>();
+
+        /// <summary>
+        /// DedicatedAsync processing thread.
+        /// </summary>
+        /// <param name="pollingDelayInMilliseconds">The await delay when there are no tasks in the queue.</param>
+        public DedicatedAsynchronousProcessingThread(int pollingDelayInMilliseconds = 50)
+        {
+            _pollingDelayInMilliseconds = pollingDelayInMilliseconds;
+        }
+
+        internal void Start()
+        {
+            ThrowIfDisposed();
+            ThrowIfAlreadyStarted();
+
+            _processingThread = Task.Factory.StartNew(
+                ProcessAsync,
+                CancellationToken.None,
+                TaskCreationOptions.LongRunning | TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default);
+        }
+
+        /// <summary>
+        /// Queueus a task for execution.
+        /// </summary>
+        /// <param name="task">Task to be executed.</param>
+        internal void Enqueue(Func<Task> task)
+        {
+            ThrowIfDisposed();
+            ThrowIfNotAlreadyStarted();
+            _taskQueue.Enqueue(task);
+        }
+
+        private async Task ProcessAsync()
+        {
+            while (!_isClosed)
+            {
+                try
+                {
+                    if (_taskQueue.TryDequeue(out var result))
+                    {
+                        await result();
+                    }
+                    else
+                    {
+                        await Task.Delay(_pollingDelayInMilliseconds);
+                    }
+                }
+                catch
+                {
```

Should we log something?

---

We don't have access to a proper logger here, and it'd be a bad practice to let the exception propagate in these asynchronous tasks cause they are not awaited.